File: algopt/rebalancer/examples/website/cookbook/affinity_placement/negative_affinity.py
#!/usr/bin/env python3
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

"""
Negative Affinity (Avoid Assignments) Variation

This variation demonstrates how to use AvoidAssignmentsSpec to enforce hard
incompatibility constraints. This is useful when certain workloads CANNOT run
on specific hosts due to hardware requirements or security restrictions.
Unlike soft affinity, this creates a hard constraint that cannot be violated.
"""

from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from rebalancer.interface.thrift.v2.ProblemSolver.thrift_types import (
    ConstraintSpecs,
    SolverSpecs,
)
from rebalancer.interface.thrift.v2.ProblemSpecs.thrift_types import (
    AvoidAssignment,
    AvoidAssignmentsSpec,
    CapacitySpec,
)
from rebalancer.interface.thrift.v2.SolverSpecs.thrift_types import (
    LocalSearchSolverSpec,
)


def place_with_negative_affinity():
    """Place services with hard incompatibility constraints."""
    solver = ProblemSolver(
        service_name="microservice-placer", service_scope="production"
    )

    solver.setObjectName("service_instance")
    solver.setContainerName("host")

    # Host hardware profiles
    host_profiles = {
        "gpu_host_0": {"gpu": True, "memory": 128},
        "gpu_host_1": {"gpu": True, "memory": 128},
        "standard_host_0": {"gpu": False, "memory": 64},
        "standard_host_1": {"gpu": False, "memory": 64},
        "standard_host_2": {"gpu": False, "memory": 64},
        "standard_host_3": {"gpu": False, "memory": 64},
    }

    # Initial assignment
    assignment = {host: [] for host in host_profiles.keys()}

    # Create GPU instances
    gpu_instances = [f"ml_training_{i}" for i in range(10)]
    cpu_instances = [f"web_server_{i}" for i in range(20)]

    # Random initial assignment (may violate requirements)
    all_hosts = list(host_profiles.keys())
    for i, inst in enumerate(gpu_instances + cpu_instances):
        host_idx = i % len(all_hosts)
        assignment[all_hosts[host_idx]].append(inst)

    solver.setAssignment(assignment)

    # Add dimensions
    resource_usage = {}
    for inst in gpu_instances:
        resource_usage[inst] = 8.0
    for inst in cpu_instances:
        resource_usage[inst] = 2.0

    solver.addObjectDimension("cpu_usage", resource_usage)

    host_cpu_capacity = {host: 64.0 for host in host_profiles.keys()}
    solver.addContainerDimension("cpu_capacity", host_cpu_capacity)

    # CPU capacity constraint
    solver.addConstraint(
        ConstraintSpecs(
            capacitySpec=CapacitySpec(
                name="cpu-capacity",
                scope="host",
                dimension="cpu_usage",
            )
        )
    )

    # variation_start
    # GPU workloads CANNOT run on non-GPU hosts (hard constraint)
    non_gpu_hosts = [
        host for host, profile in host_profiles.items() if not profile["gpu"]
    ]

    solver.addConstraint(
        ConstraintSpecs(
            avoidAssignmentsSpec=AvoidAssignmentsSpec(
                name="gpu-hardware-requirement",
                scope="host",
                assignments=[
                    AvoidAssignment(object=inst, scopeItems=non_gpu_hosts)
                    for inst in gpu_instances
                ],
            )
        )
    )
    # variation_end

    # Use Local Search solver
    solver.addSolver(
        SolverSpecs(localSearchSolverSpec=LocalSearchSolverSpec(timeLimitMs=30000))
    )

    # Solve
    solution = solver.solve()

    print(f"Solution found in {solution.profile.solveTime}ms")
    print(f"Objective value: {solution.objectiveValue}")
    print(f"Instances moved: {solution.profile.moveCount}")

    # Verify GPU placement
    print("\nGPU placement verification:")
    for host, instances in solution.assignment.items():
        is_gpu_host = host_profiles[host]["gpu"]
        gpu_count = sum(1 for inst in instances if inst in gpu_instances)
        if gpu_count > 0:
            if is_gpu_host:
                print(f"  ✅ {host}: {gpu_count} GPU instances (valid)")
            else:
                print(f"  ❌ {host}: {gpu_count} GPU instances (VIOLATION!)")

    return solution


if __name__ == "__main__":
    place_with_negative_affinity()

File: algopt/rebalancer/examples/website/cookbook/affinity_placement/negative_affinity.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

/**
 * Negative Affinity (Avoid Assignments) Variation
 *
 * This variation demonstrates how to use AvoidAssignmentsSpec to enforce hard
 * incompatibility constraints. This is useful when certain workloads CANNOT run
 * on specific hosts due to hardware requirements or security restrictions.
 * Unlike soft affinity, this creates a hard constraint that cannot be violated.
 */

#include "algopt/rebalancer/interface/ProblemSolver.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/ProblemSpecs_types.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/SolverSpecs_types.h"

#include <fmt/format.h>
#include <folly/container/irange.h>
#include <folly/executors/CPUThreadPoolExecutor.h>
#include <folly/init/Init.h>

#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace facebook::rebalancer::interface;

struct HostProfile {
  bool gpu;
  int memory;
};

int main(int argc, char* argv[]) {
  const folly::Init init(&argc, &argv);
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "microservice-placer", "production");

  solver.setObjectName("service_instance");
  solver.setContainerName("host");

  // Host hardware profiles
  std::map<std::string, HostProfile> host_profiles = {
      {"gpu_host_0", {.gpu = true, .memory = 128}},
      {"gpu_host_1", {.gpu = true, .memory = 128}},
      {"standard_host_0", {.gpu = false, .memory = 64}},
      {"standard_host_1", {.gpu = false, .memory = 64}},
      {"standard_host_2", {.gpu = false, .memory = 64}},
      {"standard_host_3", {.gpu = false, .memory = 64}},
  };

  // Initial assignment
  std::map<std::string, std::vector<std::string>> assignment;
  for (const auto& [host, _] : host_profiles) {
    assignment[host] = {};
  }

  // Create GPU instances
  std::vector<std::string> gpu_instances;
  gpu_instances.reserve(10);
  for (const auto i : folly::irange(10)) {
    gpu_instances.push_back(fmt::format("ml_training_{}", i));
  }

  std::vector<std::string> cpu_instances;
  cpu_instances.reserve(20);
  for (const auto i : folly::irange(20)) {
    cpu_instances.push_back(fmt::format("web_server_{}", i));
  }

  // Random initial assignment
  std::vector<std::string> all_hosts;
  all_hosts.reserve(host_profiles.size());
  for (const auto& [host, _] : host_profiles) {
    all_hosts.push_back(host);
  }

  std::vector<std::string> all_instances;
  all_instances.insert(
      all_instances.end(), gpu_instances.begin(), gpu_instances.end());
  all_instances.insert(
      all_instances.end(), cpu_instances.begin(), cpu_instances.end());

  for (const auto i : folly::irange(all_instances.size())) {
    int host_idx = i % all_hosts.size();
    assignment[all_hosts[host_idx]].push_back(all_instances[i]);
  }

  solver.setAssignment(assignment);

  // Add dimensions
  std::map<std::string, double> resource_usage;
  for (const auto& inst : gpu_instances) {
    resource_usage[inst] = 8.0;
  }
  for (const auto& inst : cpu_instances) {
    resource_usage[inst] = 2.0;
  }
  solver.addObjectDimension("cpu_usage", resource_usage);

  std::map<std::string, double> host_cpu_capacity;
  for (const auto& [host, _] : host_profiles) {
    host_cpu_capacity[host] = 64.0;
  }
  solver.addContainerDimension("cpu_capacity", host_cpu_capacity);

  // CPU capacity constraint
  CapacitySpec cpuCapacity;
  cpuCapacity.name() = "cpu-capacity";
  cpuCapacity.scope() = "host";
  cpuCapacity.dimension() = "cpu_usage";
  solver.addConstraint(cpuCapacity);

  // variation_start
  // GPU workloads CANNOT run on non-GPU hosts
  std::vector<std::string> non_gpu_hosts;
  for (const auto& [host, profile] : host_profiles) {
    if (!profile.gpu) {
      non_gpu_hosts.push_back(host);
    }
  }

  std::vector<AvoidAssignment> avoidances;
  for (const auto& inst : gpu_instances) {
    AvoidAssignment avoid;
    avoid.object() = inst;
    avoid.scopeItems() = non_gpu_hosts;
    avoidances.push_back(avoid);
  }

  AvoidAssignmentsSpec gpuRequirement;
  gpuRequirement.name() = "gpu-hardware-requirement";
  gpuRequirement.scope() = "host";
  gpuRequirement.assignments() = avoidances;
  solver.addConstraint(gpuRequirement);
  // variation_end

  // Use Local Search solver
  LocalSearchSolverSpec localSearch;
  localSearch.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  localSearch.solveTime() = 30000;
  solver.addSolver(localSearch);

  // Solve
  auto solution = solver.solve();

  std::cout << "AssignmentSolution found in " << "<solve time>" << "ms\n";
  std::cout << "Objective value: " << *solution.finalObjective()->value()
            << "\n";
  std::cout << "Instances moved: " << "<move count>" << "\n";

  return 0;
}
